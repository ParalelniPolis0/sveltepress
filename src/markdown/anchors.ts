import type { Plugin } from 'unified'
import { visit } from 'unist-util-visit'

const anchors: Plugin<any[], any> = () => {
  return (tree, vFile) => {
    const anchors = []
    visit(tree, (node, idx, parent) => {
      if (node.type === 'heading' && !node.data?.anchorAdded) {
        if (!node.data) {
          node.data = {
            anchorAdded: true,
          }
        }
        else {
          node.data.anchorAdded = true
        }
        const title = node.children.filter(c => ['text', 'inlineCode'].includes(c.type)).map(c => c.value).join('')

        parent.children.splice(idx, 1, {
          type: 'slugId',
          data: {
            hName: 'div',
            hProperties: {
              id: title,
              className: 'svp-anchor-item',
            },
          },
        }, node)
        anchors.push({
          slugId: title,
          title,
          depth: node.depth,
        })
      }
    })
    vFile.data.anchors = anchors
  }
}

export default anchors
