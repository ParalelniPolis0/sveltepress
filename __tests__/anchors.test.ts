import { describe, expect, it } from 'vitest'
import { compile } from 'mdsvex'
import anchors from '../src/markdown/anchors'

const md = `
## Foo \`foo\`

[foo](https://www.google.com/)
some content

### Bar

some other content

`

describe('anchors', () => {
  it('simple', async () => {
    const r = await compile(md, {
      remarkPlugins: [anchors, () => (tree) => {
        expect(JSON.stringify(tree, null, 2)).toMatchSnapshot()
      }],
    })
    expect(r).toMatchInlineSnapshot(`
      {
        "code": "
      <div id=\\"Foo foo\\" class=\\"svp-anchor-item\\"></div>
      <h2>Foo <code>foo</code></h2>
      <p><a href=\\"https://www.google.com/\\" rel=\\"nofollow\\">foo</a>
      some content</p>
      <div id=\\"Bar\\" class=\\"svp-anchor-item\\"></div>
      <h3>Bar</h3>
      <p>some other content</p>
      ",
        "data": {
          "anchors": [
            {
              "depth": 2,
              "slugId": "Foo foo",
              "title": "Foo foo",
            },
            {
              "depth": 3,
              "slugId": "Bar",
              "title": "Bar",
            },
          ],
        },
        "map": "",
      }
    `)
  })
})
