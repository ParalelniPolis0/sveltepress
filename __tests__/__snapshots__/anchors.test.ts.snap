// Vitest Snapshot v1

exports[`anchors > simple 1`] = `
"{
  \\"type\\": \\"root\\",
  \\"children\\": [
    {
      \\"type\\": \\"slugId\\",
      \\"data\\": {
        \\"hName\\": \\"div\\",
        \\"hProperties\\": {
          \\"id\\": \\"Foo foo\\",
          \\"className\\": \\"svp-anchor-item\\"
        }
      }
    },
    {
      \\"type\\": \\"heading\\",
      \\"depth\\": 2,
      \\"children\\": [
        {
          \\"type\\": \\"text\\",
          \\"value\\": \\"Foo \\",
          \\"position\\": {
            \\"start\\": {
              \\"line\\": 2,
              \\"column\\": 4,
              \\"offset\\": 4
            },
            \\"end\\": {
              \\"line\\": 2,
              \\"column\\": 8,
              \\"offset\\": 8
            },
            \\"indent\\": []
          }
        },
        {
          \\"type\\": \\"inlineCode\\",
          \\"value\\": \\"foo\\",
          \\"position\\": {
            \\"start\\": {
              \\"line\\": 2,
              \\"column\\": 8,
              \\"offset\\": 8
            },
            \\"end\\": {
              \\"line\\": 2,
              \\"column\\": 13,
              \\"offset\\": 13
            },
            \\"indent\\": []
          }
        }
      ],
      \\"position\\": {
        \\"start\\": {
          \\"line\\": 2,
          \\"column\\": 1,
          \\"offset\\": 1
        },
        \\"end\\": {
          \\"line\\": 2,
          \\"column\\": 13,
          \\"offset\\": 13
        },
        \\"indent\\": []
      },
      \\"data\\": {
        \\"anchorAdded\\": true
      }
    },
    {
      \\"type\\": \\"paragraph\\",
      \\"children\\": [
        {
          \\"type\\": \\"link\\",
          \\"title\\": null,
          \\"url\\": \\"https://www.google.com/\\",
          \\"children\\": [
            {
              \\"type\\": \\"text\\",
              \\"value\\": \\"foo\\",
              \\"position\\": {
                \\"start\\": {
                  \\"line\\": 4,
                  \\"column\\": 2,
                  \\"offset\\": 16
                },
                \\"end\\": {
                  \\"line\\": 4,
                  \\"column\\": 5,
                  \\"offset\\": 19
                },
                \\"indent\\": []
              }
            }
          ],
          \\"position\\": {
            \\"start\\": {
              \\"line\\": 4,
              \\"column\\": 1,
              \\"offset\\": 15
            },
            \\"end\\": {
              \\"line\\": 4,
              \\"column\\": 31,
              \\"offset\\": 45
            },
            \\"indent\\": []
          },
          \\"data\\": {
            \\"hProperties\\": {
              \\"rel\\": [
                \\"nofollow\\"
              ]
            }
          }
        },
        {
          \\"type\\": \\"text\\",
          \\"value\\": \\"\\\\nsome content\\",
          \\"position\\": {
            \\"start\\": {
              \\"line\\": 4,
              \\"column\\": 31,
              \\"offset\\": 45
            },
            \\"end\\": {
              \\"line\\": 5,
              \\"column\\": 13,
              \\"offset\\": 58
            },
            \\"indent\\": [
              1
            ]
          }
        }
      ],
      \\"position\\": {
        \\"start\\": {
          \\"line\\": 4,
          \\"column\\": 1,
          \\"offset\\": 15
        },
        \\"end\\": {
          \\"line\\": 5,
          \\"column\\": 13,
          \\"offset\\": 58
        },
        \\"indent\\": [
          1
        ]
      }
    },
    {
      \\"type\\": \\"slugId\\",
      \\"data\\": {
        \\"hName\\": \\"div\\",
        \\"hProperties\\": {
          \\"id\\": \\"Bar\\",
          \\"className\\": \\"svp-anchor-item\\"
        }
      }
    },
    {
      \\"type\\": \\"heading\\",
      \\"depth\\": 3,
      \\"children\\": [
        {
          \\"type\\": \\"text\\",
          \\"value\\": \\"Bar\\",
          \\"position\\": {
            \\"start\\": {
              \\"line\\": 7,
              \\"column\\": 5,
              \\"offset\\": 64
            },
            \\"end\\": {
              \\"line\\": 7,
              \\"column\\": 8,
              \\"offset\\": 67
            },
            \\"indent\\": []
          }
        }
      ],
      \\"position\\": {
        \\"start\\": {
          \\"line\\": 7,
          \\"column\\": 1,
          \\"offset\\": 60
        },
        \\"end\\": {
          \\"line\\": 7,
          \\"column\\": 8,
          \\"offset\\": 67
        },
        \\"indent\\": []
      },
      \\"data\\": {
        \\"anchorAdded\\": true
      }
    },
    {
      \\"type\\": \\"paragraph\\",
      \\"children\\": [
        {
          \\"type\\": \\"text\\",
          \\"value\\": \\"some other content\\",
          \\"position\\": {
            \\"start\\": {
              \\"line\\": 9,
              \\"column\\": 1,
              \\"offset\\": 69
            },
            \\"end\\": {
              \\"line\\": 9,
              \\"column\\": 19,
              \\"offset\\": 87
            },
            \\"indent\\": []
          }
        }
      ],
      \\"position\\": {
        \\"start\\": {
          \\"line\\": 9,
          \\"column\\": 1,
          \\"offset\\": 69
        },
        \\"end\\": {
          \\"line\\": 9,
          \\"column\\": 19,
          \\"offset\\": 87
        },
        \\"indent\\": []
      }
    }
  ],
  \\"position\\": {
    \\"start\\": {
      \\"line\\": 1,
      \\"column\\": 1,
      \\"offset\\": 0
    },
    \\"end\\": {
      \\"line\\": 11,
      \\"column\\": 1,
      \\"offset\\": 89
    }
  }
}"
`;
